### PR TITLE
Optimize marshalling codepath in networks check

### PR DESCRIPTION
**Pull request by leeavital**
This commit brings in a version of the agent payload which uses [protoc-gen-gostreamer](https://github.com/leeavital/protoc-gen-gostreamer/).

This can avoid many intermediate allocations when
serializing network connections over the system-probe UDS. For background, when the system-probe responds to a /connection RPC, the following happens:

1. connections stored in state.go are pulled out as a network.Connections which under the hood is a big list of []network.Connection
2. the encoding package converts this []network.Connection to model.Connections, which under the hood is a []model.Connection with several buffers that hold compact representations of tags and DNS data as `[]byte`s
3. Then we call proto.Marshal on the model.Connection object
4. ...and write that into a response writer

In terms of allocations, step 2 creates a lot of short lived small objects since each network.Connection needs to be converted to a model.Connection. Alongside with a number of larger allocations for DNS and tag sets. Before this PR, formatConnection incurred an allocation per-connection each time an RPC happened.

Step 3 does one *big* allocation since the whole model.Connections object needs to be allocated in a contiguous block of memory. Before this PR, this happened in the Marshal method in protoSerializer.

The new code generated for protobuf marshalling allows different parts of the network.Connections object to be written in separate calls. I.e. instead of converting each network.Connection to model.Connection object, each field in network.Connection can be written to the protobuf separately without incurring a heap allocation.

This means a lot of encoding code needs to be re-written from looking like this:

```
func FormatFoo(f *network.Foo) *model.Foo {
  return &model.Foo{
    x: int32(f.x),
  }
}
```

To

```
func FormatFoo(f *network.Foo, builder *model.FooBuilder) {
  builder.SetX(int32(f.x))
}
```

See FormatConnections for the most comprehensive example of how this is done.

[Benchmarks](https://ddstaging.datadoghq.com/notebook/5983918/perftest-fast-encode?range=1800000&start=1689038044502&live=false) show a 36% memory improvement in the best case and 28% in the average case in a load test. CPU impact appears to be negligible.

image

image

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/security/ebpf/compile.go [hunk 1]
```diff
@@ -18,7 +18,6 @@ import (
 // TODO change probe.c path to runtime-compilation specific version
 //go:generate $GOPATH/bin/include_headers pkg/security/ebpf/c/prebuilt/probe.c pkg/ebpf/bytecode/build/runtime/runtime-security.c pkg/security/ebpf/c/include pkg/ebpf/c
 //go:generate $GOPATH/bin/integrity pkg/ebpf/bytecode/build/runtime/runtime-security.c pkg/ebpf/bytecode/runtime/runtime-security.go runtime
-//go:generate go run github.com/DataDog/datadog-agent/pkg/security/secl/model/bpf_maps_generator -runtime-path ../../ebpf/bytecode/build/runtime/runtime-security.c -output ../../security/secl/model/consts_map_names.go -pkg-name model
```
Comment (paulcacheux) [inline pkg/security/ebpf/compile.go:21]: Why was this removed ?

**Review by paulcacheux — CHANGES_REQUESTED** [review]

File: go.mod [hunk 2]
```diff
@@ -560,6 +559,7 @@ require (
 require github.com/lorenzosaino/go-sysctl v0.3.1
 
 require (
+	github.com/DataDog/agent-payload/v5 v5.0.94-0.20230821121105-8311d9ecf0bf
```
Comment (leeavital) [inline go.mod:562]: TODO: use a tagged version

File: pkg/security/ebpf/compile.go (hunk 1 above)
Comment (leeavital) [reply pkg/security/ebpf/compile.go]: Good question, I think this was an accident. It's restored now!

**Review by paulcacheux — APPROVED** [review]

File: pkg/network/encoding/format.go [hunk 3]
```diff
@@ -125,46 +150,26 @@ func FormatConnectionTelemetry(tel map[network.ConnTelemetryType]int64) map[stri
...
+			w.SetKey(k)
+			w.SetValue(v)
+		})
 	}
 
-	return ret
 }
 
-func FormatCORETelemetry(telByAsset map[string]int32) map[string]model.COREResult {
+func FormatCORETelemetry(builder *model.ConnectionsBuilder, telByAsset map[string]int32) {
```
Comment (akarpz) [inline pkg/network/encoding/format.go:162]: looks like CI is failing here: 
```pkg/network/encoding/format.go:162:1: exported: exported function FormatCORETelemetry should have comment or be unexported```

File: pkg/network/encoding/encoding.go [hunk 4]
```diff
@@ -80,27 +80,46 @@ func modelConnections(conns *network.Connections) *model.Connections {
 	dnsFormatter := newDNSFormatter(conns, ipc)
 	tagsSet := network.NewTagsSet()
 
-	for i, conn := range conns.Conns {
-		agentConns[i] = FormatConnection(conn, routeIndex, httpEncoder, http2Encoder, kafkaEncoder, dnsFormatter, ipc, tagsSet)
+	for _, conn := range conns.Conns {
+		builder.AddConns(func(builder *model.ConnectionBuilder) {
+			FormatConnection(builder, conn, routeIndex, httpEncoder, http2Encoder, kafkaEncoder, dnsFormatter, ipc, tagsSet)
+		})
 	}
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:87]: seems like a classic golang-gotcha

File: pkg/network/encoding/encoding.go [hunk 5]
```diff
@@ -80,27 +80,46 @@ func modelConnections(conns *network.Connections) *model.Connections {
...
+
+	dnsFormatter.FormatDNS(builder)
+
+	for _, tag := range tagsSet.GetStrings() {
+		builder.AddTags(tag)
+	}
+
+	FormatConnectionTelemetry(builder, conns.ConnTelemetry)
+	FormatCompilationTelemetry(builder, conns.CompilationTelemetryByAsset)
+	FormatCORETelemetry(builder, conns.CORETelemetryByAsset)
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:119]: does it make sense to export those functions?

File: pkg/network/encoding/format.go [hunk 6]
```diff
@@ -45,78 +39,109 @@ func (ipc ipCache) Get(addr util.Address) string {
...
+
+	builder.SetLaddr(func(w *model.AddrBuilder) {
+		w.SetIp(ipc.Get(conn.Source))
+		w.SetPort(int32(conn.SPort))
+		w.SetContainerId(containerID)
+	})
+	builder.SetRaddr(func(w *model.AddrBuilder) {
+		w.SetIp(ipc.Get(conn.Dest))
+		w.SetPort(int32(conn.DPort))
+	})
```
Comment (guyarb) [inline pkg/network/encoding/format.go:59]: you must check conn.Source and conn.Dest are not nil, otherwise `ipc.Get` will panic

File: pkg/network/encoding/format.go [hunk 7]
```diff
@@ -45,78 +39,109 @@ func (ipc ipCache) Get(addr util.Address) string {
...
 		containerID = *conn.ContainerID
 	}
-	c.Laddr = formatAddr(conn.Source, conn.SPort, containerID, ipc)
-	c.Raddr = formatAddr(conn.Dest, conn.DPort, "", ipc)
-	c.Family = formatFamily(conn.Family)
-	c.Type = formatType(conn.Type)
-	c.IsLocalPortEphemeral = formatEphemeralType(conn.SPortIsEphemeral)
-	c.LastBytesSent = conn.Last.SentBytes
-	c.LastBytesReceived = conn.Last.RecvBytes
-	c.LastPacketsSent = conn.Last.SentPackets
```
Comment (guyarb) [inline pkg/network/encoding/format.go:71]: this is missing in the new format

File: pkg/network/encoding/format.go [hunk 8]
```diff
@@ -45,78 +39,109 @@ func (ipc ipCache) Get(addr util.Address) string {
...
+	builder.SetFamily(uint64(formatFamily(conn.Family)))
+	builder.SetType(uint64(formatType(conn.Type)))
+	builder.SetIsLocalPortEphemeral(uint64(formatEphemeralType(conn.SPortIsEphemeral)))
+	builder.SetLastBytesSent(conn.Last.SentBytes)
+	builder.SetLastBytesReceived(conn.Last.RecvBytes)
+	builder.SetLastRetransmits(conn.Last.Retransmits)
+	builder.SetLastPacketsReceived(conn.Last.RecvPackets)
+	builder.SetDirection(uint64(formatDirection(conn.Direction)))
+	builder.SetNetNS(conn.NetNS)
+	builder.SetRemoteNetworkId("")
```
Comment (guyarb) [inline pkg/network/encoding/format.go:69]: do we need to explicitly set it if it is empty / default value?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: go.mod (hunk 2 above)
[the remark this answers is not in the document]
Comment (akarpz) [reply go.mod:562]: this was likely using the commit hash of the branch before it was merged and needs to be changed now that https://github.com/DataDog/agent-payload/pull/264 this is merged

File: releasenotes/notes/sysprobe-optimize-proto-97f335159188955f.yaml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# Each section from every release note are combined when the
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - Improve marshalling performance in the system-probe -> process-agent path. This improves memory footprint when NPM and/or USM are enabled.
```
Comment (hestonhoffman) [inline releasenotes/notes/sysprobe-optimize-proto-97f335159188955f.yaml:10]:
```suggestion
  - Improve marshalling performance in the ``system-probe`` -> ``process-agent`` path. This improves memory footprint when NPM and/or USM are enabled.
```

**Review by hestonhoffman — APPROVED** [review]
Small release note edit

File: pkg/network/encoding/json.go [hunk 10]
```diff
@@ -21,12 +22,16 @@ type jsonSerializer struct {
...
+func (j jsonSerializer) Marshal(conns *network.Connections, writer io.Writer) error {
+	out := bytes.NewBuffer(nil)
+	modelConnections(model.NewConnectionsBuilder(out), conns)
+
+	var payload model.Connections
+	if err := payload.Unmarshal(out.Bytes()); err != nil {
+		return err
+	}
+
+	return j.marshaller.Marshal(writer, &payload)
```
[the remark this answers is not in the document]
Comment (leeavital) [reply pkg/network/encoding/json.go:34]: gostreamer does not have a way to emit json, so the way to get json is to go `network.Connections -> proto bytes -> model.Connection -> json (via jsonpb)`

Given this endpoint is only used for debugging purposes, I wasn't too concerned about the performance implications.

File: pkg/network/encoding/encoding.go (hunk 5 above)
Comment (leeavital) [reply pkg/network/encoding/encoding.go:119]: These were already public, and I was keeping the diff as small as possible, but if you think it's useful, I think these can be made private

File: pkg/network/encoding/format.go (hunk 6 above)
Comment (leeavital) [reply pkg/network/encoding/format.go:59]: Are there instances where source/dest are nil?  IMO those should be filtered out before we get to encoding.

This worked fine on the testing environment, and passes unit tests, so if it's a possibility that's a pointer to us missing some tests.

File: pkg/network/encoding/format.go (hunk 8 above)
Comment (leeavital) [reply pkg/network/encoding/format.go:69]: We shouldn't no

File: pkg/network/encoding/json.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/encoding/json.go:34]: While that's true, I think we should still measure the impact of the new implementation on the debug endpoint, as for example if it suddenly consumes 100MB, it might lead to OOMs (staging/production/customer) and lost of precious debugging information

File: pkg/network/encoding/encoding.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/encoding/encoding.go:119]: The diff showed it as newly exported, but you're right
I'm fine with keeping it as it

File: pkg/network/encoding/format.go (hunk 6 above)
Comment (guyarb) [reply pkg/network/encoding/format.go:59]: the original formatAddr checked it, I'm not against of filtering it out, I'm only concern that the code "was there for a reason"

Comment (guyarb) [reply pkg/network/encoding/format.go:59]: [seeing](https://github.com/DataDog/datadog-agent/pull/3783) @sunhay as the original author 4 years ago

Comment (leeavital) [reply pkg/network/encoding/format.go:59]: Thanks for hunting that down. 

Given the age of the code, and that this ran on several internal clusters already, I'll leave the code as is. Let me know if you disagree.

**Review by guyarb — APPROVED** [review]
Great work!

**Review by ogaca-dd — APPROVED** [review]
LGTM for files owned by agent shared components.

**Review by hmahmood — APPROVED** [review]

**Review by akarpz — APPROVED** [review]
